lib/Conversion/TorchToLinalg/Linear.cpp

```diff
@@ -448,6 +449,242 @@ class ConvertAtenLinearOp : public OpConversionPattern<AtenLinearOp> {
+    }
+    SmallVector<int64_t> strideInts;
+    if (!matchPattern(op.stride(), m_TorchConstantIntList(strideInts)))
+      return rewriter.notifyMatchFailure(op,
+                                         "only support constant int strides");
+    SmallVector<int64_t> dilationInts;
+    if (!matchPattern(op.dilation(), m_TorchConstantIntList(dilationInts)))
+      return rewriter.notifyMatchFailure(op,
+                                         "only support constant int dilations");
+
+    for (size_t i = 0; i < paddingInts.size(); i++) {
+      if (paddingInts[i] != 0)
+        return rewriter.notifyMatchFailure(op, "only support 0 padding");
+      paddingInts[i] += 1;
+    }
+
+    SmallVector<Value> paddingIntValues =
+        getAsConstantIntValues(rewriter, loc, paddingInts);
+    SmallVector<Value> dilationIntValues =
+        getAsConstantIntValues(rewriter, loc, dilationInts);
+    SmallVector<Value> strideIntValues =
+        getAsConstantIntValues(rewriter, loc, strideInts);
+
+    auto stridesAttr = rewriter.getI64VectorAttr(strideInts);
+    auto dilationAttr = rewriter.getI64VectorAttr(dilationInts);
+
+    Value c1 =
+        rewriter.create<arith::ConstantOp>(loc, IntegerAttr::get(intType, 1));
+    Value c0 =
+        rewriter.create<arith::ConstantOp>(loc, FloatAttr::get(elementType, 0));
+    Value ci0 =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getI64IntegerAttr(0));
+    Value groupEqual1 = rewriter.create<arith::CmpIOp>(
+        loc, arith::CmpIPredicate::eq, groups, c1);
+    rewriter.create<cf::AssertOp>(
+        loc, groupEqual1, rewriter.getStringAttr("expect groups to be 1"));
+
+    // Pad the input tensor according to padding.
+    SmallVector<int64_t> paddingIncludingNC = {0, 0};
+    paddingIncludingNC.insert(paddingIncludingNC.end(), paddingInts.begin(),
+                              paddingInts.end());
+
+    SmallVector<Value> gradDims, inputDims, weightDims;
+    for (size_t i = 0; i < inRank; i++) {
+      gradDims.push_back(getDimOp(rewriter, loc, grad, i));
+      inputDims.push_back(getDimOp(rewriter, loc, input, i));
+      weightDims.push_back(getDimOp(rewriter, loc, weight, i));
+    }
+
+    auto getConvOutputDim = [&](Value lhDim, Value rhDim, Value padding,
+                                Value stride, Value dilation,
+                                bool valid) -> Value {
+      if (valid) {
+        return torch_to_linalg::getOutputDimForConvOps(
+            rewriter, loc, lhDim, ci0, dilation, castIndexToInt(rhDim), stride);
+      }
+      // TODO: dilation, stride
+      Value c1 =
+          rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(1));
+      Value res = rewriter.create<arith::AddIOp>(loc, lhDim, rhDim);
+      return rewriter.create<arith::SubIOp>(loc, res, c1);
+    };
+    auto convolve = [&](Value lhs, Value rhs, SmallVector<Value> &lhDims,
+                        SmallVector<Value> &rhDims, bool valid) -> Value {
```

Comment:
what does `valid` represent here?

Reply:
Whether the convolution is valid (i.e. unpadded)